File: src/utils/startup.ts
import { parse as parseYaml } from "yaml";
import type { Mode } from "../store/useIdeStore";

export type StartupAction =
  | { action: "open_files"; files: string[] }
  | { action: "set_mode"; mode: Mode }
  | {
      action: "run_terminal";
      commands: { title: string; cmd: string }[];
    };

export interface StartupConfig {
  startup: StartupAction[];
}

export function parseStartup(yaml: string): StartupConfig {
  const parsed = parseYaml(yaml);
  if (
    !parsed ||
    typeof parsed !== "object" ||
    !Array.isArray((parsed as { startup?: unknown }).startup)
  ) {
    throw new Error("invalid .ide-startup.yaml: missing 'startup' list");
  }
  const config = parsed as StartupConfig;
  for (const step of config.startup) {
    if (!step || typeof step !== "object" || !("action" in step)) {
      throw new Error("invalid startup step: missing 'action'");
    }
    if (
      step.action !== "open_files" &&
      step.action !== "set_mode" &&
      step.action !== "run_terminal"
    ) {
      throw new Error(`unknown startup action: ${(step as { action: string }).action}`);
    }
  }
  return config;
}

export interface StartupRunner {
  setMode: (mode: Mode) => void;
  openFile: (path: string) => void;
  addTerminal: (title: string, initialCmd?: string) => string;
}

export function applyStartup(
  config: StartupConfig,
  runner: StartupRunner
): void {
  for (const step of config.startup) {
    switch (step.action) {
      case "set_mode":
        runner.setMode(step.mode);
        break;
      case "open_files":
        step.files.forEach((f) => runner.openFile(f));
        break;
      case "run_terminal":
        step.commands.forEach((c) => runner.addTerminal(c.title, c.cmd));
        break;
    }
  }
}

File: src-tauri/src/startup.rs
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum StartupAction {
    OpenFiles { files: Vec<String> },
    SetMode { mode: String },
    RunTerminal { commands: Vec<RunTerminalCommand> },
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RunTerminalCommand {
    pub title: String,
    pub cmd: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct StartupConfig {
    pub startup: Vec<StartupAction>,
}

fn default_paths() -> Vec<PathBuf> {
    let mut paths = Vec::new();
    if let Ok(cwd) = std::env::current_dir() {
        paths.push(cwd.join(".ide-startup.yaml"));
        if let Some(parent) = cwd.parent() {
            paths.push(parent.join(".ide-startup.yaml"));
        }
    }
    paths
}

#[tauri::command]
pub fn read_startup_config() -> Result<Option<StartupConfig>, String> {
    for path in default_paths() {
        if path.is_file() {
            let raw = std::fs::read_to_string(&path)
                .map_err(|e| format!("failed to read {}: {e}", path.display()))?;
            let cfg: StartupConfig = serde_yaml::from_str(&raw)
                .map_err(|e| format!("invalid yaml in {}: {e}", path.display()))?;
            return Ok(Some(cfg));
        }
    }
    Ok(None)
}
